cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/cloudstorage/coordinated/MultiClusterContainer.java

```diff
@@ -0,0 +1,190 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.cassandra.spark.bulkwriter.cloudstorage.coordinated;
+
+import java.io.Serializable;
+import java.util.Map;
+import java.util.NoSuchElementException;
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.BiConsumer;
+import java.util.function.UnaryOperator;
+
+import com.google.common.base.Preconditions;
+
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * A container to hold values per cluster. Each value is identified by its belonging cluster id.
+ * It is compatible with single cluster, i.e. no cluster id is defined for the value. Pass null clusterId to get the value.
+ * However, the container does not permit holding values for both multi-cluster case and single cluster case;
+ * otherwise, {@link IllegalStateException} is thrown.
+ * @param <T> value type
+ */
+public class MultiClusterContainer<T> implements Serializable, MultiClusterSupport<T>
+{
+    private static final long serialVersionUID = 8387168256773834417L;
+    // A special (non-string) key used by non-coordinated write; The actual value (Integer) is serializable constant to ensure ser/deser stability
+    private static final Object SENTINEL_KEY = Integer.MAX_VALUE;
+
+    // For coordinated write, the key should be String values of clusterId
+    private final Map<Object, T> byCluster = new ConcurrentHashMap<>();
+
+    @Nullable
+    @Override
+    public T getValueOrNull(@Nullable String clusterId)
+    {
+        return byCluster.get(key(clusterId));
+    }
+
+    /**
+     * @return any value in the container. Null is returned if the container is empty.
+     */
+    @Nullable
+    public T getAnyValue()
+    {
+        T v = null;
+        if (!byCluster.isEmpty())
```

Comment:
Now that we have the whole sentinel value for null, is there any case where `byCluster` would be empty? Wouldn't you always at least have the `null` key for non-coordinated writes? If so, we should be able to get rid of most/all of these checks.

Reply:
The value container can be empty, when no entries are populated. Empty check is still required to get any value from the map (, since it returns the first value from `values()`).